crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs

```diff
@@ -1189,6 +1194,110 @@ impl Drop for IndexeddbCryptoStore {
     }
 }
 
+/// Fetch items from an object store in batches, transform each item using
+/// the supplied function, and stuff the transformed items into a single
+/// vector to return.
+async fn fetch_from_object_store_batched<R, F>(
+    object_store: IdbObjectStore<'_>,
+    f: F,
+    batch_size: usize,
+) -> Result<Vec<R>>
+where
+    F: Fn(JsValue) -> Result<R>,
+{
+    enum NextBatch {
+        First,          // In the first batch we don't start from a key
+        Later(JsValue), // In later batches we have a starting key
+    }
+
+    let mut result = Vec::new();
+    let mut next_batch = NextBatch::First;
+    let mut batch_n = 0;
+
+    loop {
+        debug!("Fetching Indexed DB records starting from {}", batch_n * batch_size);
+
+        // See https://github.com/Alorel/rust-indexed-db/issues/31 - we
+        // would like to use `get_all_with_key_and_limit` if it ever exists
+        // but for now we use a cursor and manually limit batch size.
+
+        // Get hold of a cursor for this batch
+        let cursor = match &next_batch {
+            NextBatch::First => object_store.open_cursor(),
+
+            NextBatch::Later(latest_key) => {
+                // We want a cursor starting after the last key we processed
+                let after_latest_key = IdbKeyRange::lower_bound_with_open(latest_key, true);
+
+                // This key should always be valid since we got hold of it
+                // within the same transaction.
+                let after_latest_key = after_latest_key.expect("Key was not valid!");
+
+                object_store.open_cursor_with_range(&after_latest_key)
+            }
+        }?
+        .await?;
+
+        // Fetch batch_size records into result
+        let next_key = fetch_batch(cursor, batch_size, &f, &mut result).await?;
+        if let Some(next_key) = next_key {
+            next_batch = NextBatch::Later(next_key);
+        } else {
+            break;
+        }
+
+        batch_n += 1;
+    }
+
+    Ok(result)
+}
+
+/// Fetch batch_size records from the supplied cursor,
+/// and return the last key we processed, or None if
+/// we reached the end of the cursor.
+async fn fetch_batch<R, F, Q>(
+    cursor: Option<IdbCursorWithValue<'_, Q>>,
```

Comment:
light suggestion (feel free to dismiss): Since the cursor is immediately safely unwrapped in the first line of this function, and there's only one caller, I would find it makes more sense that this function takes an actual `IdbCursorWithValue` (not optional), and that the check be performed by the caller.

Reply:
Since `open_cursor_with_range` can return `None` to indicate that there are no records, it feels natural to handle this case inside `fetch_batch`. Plus, I need to handle other cases where `fetch_batch` returns `None`, so the calling code would gain an extra `if`, so I think I'll leave it as is, thanks!